api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java

```diff
@@ -94,6 +97,34 @@ public void syncBillingProjectStatus() {
     }
   }
 
+  public BillingProjectBufferEntry assignBillingProject(User user) {
+    BillingProjectBufferEntry entry = consumeBufferEntryForAssignment();
+
+    fireCloudService.addUserToBillingProject(user.getEmail(), entry.getFireCloudProjectName());
+    entry.setStatusEnum(ASSIGNED);
+    entry.setAssignedUser(user);
+    billingProjectBufferEntryDao.save(entry);
+
+    return entry;
+  }
+
+  private BillingProjectBufferEntry consumeBufferEntryForAssignment() {
+    while (billingProjectBufferEntryDao.acquireAssigningLock() != 1) {}
+
+    BillingProjectBufferEntry entry;
+    try {
+      entry = billingProjectBufferEntryDao
+          .findFirstByStatusOrderByCreationTimeAsc(
+              StorageEnums.billingProjectBufferStatusToStorage(AVAILABLE));
+      entry.setStatusEnum(ASSIGNING);
+      billingProjectBufferEntryDao.save(entry);
+    } finally {
+      billingProjectBufferEntryDao.releaseAssigningLock();
```

Comment:
What are the properties of the lock, e.g. if this call fails, or the process somehow crashes before this line can execute?

Reply:
If the process is alive, it should be able to reach this line even if an exception is thrown since its within a `finally` block. If the process crashes, MySQL should let go of the lock since the session is dead.

I did not think of the case where the release call just fails. I can add a retry loop like the one above that keeps getting called until it succeeds.

maybe with a max retry counter